### PR TITLE
fixed updates of excluded prefixes

### PR DESCRIPTION
**Pull request by Mixaster995**
Signed-off-by: Mikhail Avramenko <avramenkomihail15@gmail.com>

## Description
Excluded prefixes never removed, only appended. It's not desired behaviour.

## Issue link
closes https://github.com/networkservicemesh/sdk/issues/1179

## How Has This Been Tested?

- [x] Added unit testing to cover
- [x] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [x] Bug fix
- [ ] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/excludedprefixes/server.go [hunk 1]
```diff
@@ -37,10 +37,11 @@ import (
...
 type excludedPrefixesServer struct {
-	ctx        context.Context
-	prefixPool atomic.Value
-	once       sync.Once
-	configPath string
+	ctx              context.Context
+	prefixPool       atomic.Value
+	once             sync.Once
+	configPath       string
+	previousPrefixes map[string]struct{}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/server.go:44]: Could you explain why is this not per connection?

Comment (Mixaster995) [reply pkg/networkservice/common/excludedprefixes/server.go]: ok, fixed

**Review by glazychev-art — APPROVED** [review]

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]
as far as I can see these changes are still **not** resistant to the following scenarios:

1. Client changed his prefixes on refresh.
2. Client's and server's prefixes have a collision on the first request and do not have a collision on the next request. request1: {clientPrefixes: a, serverPrefixes:a, finalPrefixes: a}, Actual: request2: {clientPrefixes: a, serverPrefixes:b, finalPrefixes: b}. 
Expected: request2: {clientPrefixes: a, serverPrefixes:b, finalPrefixes: ab}.

**Comment by Mixaster995** [comment]
@denis-tingaikin reworked some logic, added unit test for describe scenario - please, take a look

File: pkg/networkservice/common/excludedprefixes/server.go [hunk 2]
```diff
@@ -87,15 +88,56 @@ func (eps *excludedPrefixesServer) Request(ctx context.Context, request *network
...
+
+	prefixesInfo.previousDiff = Subtract(clientPrefixes, prefixesFromFile)
+
+	log.FromContext(ctx).
+		WithField("excludedPrefixes", "server").
+		WithField("prefixesFromFile", prefixesFromFile).
+		WithField("previousPrefixesInfo", prefixesInfo).
+		Debugf("adding excluded prefixes to connection")
+
+	store(ctx, prefixesInfo)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/server.go:135]: it starts to look more and more complicated.

I think we should consider the problem on the client(nsc) side within exludeprefix.Client chain element that will simply translate response prefixes to client prefixes. In this case all applications in the chain of request can always append prefixes and don't care about diff/conflict cases.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/common/excludedprefixes/server.go (hunk 2 above)
Comment (Mixaster995) [reply pkg/networkservice/common/excludedprefixes/server.go]: ok, it sounds like a good solution, added it

**Comment by denis-tingaikin** [comment]
@Mixaster995 Could you rebase this?

**Comment by Mixaster995** [comment]
@denis-tingaikin ok, done

File: pkg/networkservice/common/excludedprefixes/utils.go [hunk 3]
```diff
@@ -50,3 +55,17 @@ func exclude(source, exclude []string) []string {
...
+	s2copy := make([]string, len(s2))
+
+	copy(s1copy, s1)
+	copy(s2copy, s2)
+
+	sort.Strings(s1copy)
+	sort.Strings(s2copy)
+
+	return reflect.DeepEqual(s1copy, s2copy)
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/utils.go:71]: Consider using github.com/google/go-cmp/cmp

We already have dep on it https://github.com/networkservicemesh/sdk/blob/main/go.mod#L17

```suggestion
```

**Review by denis-tingaikin — APPROVED** [review]
Looks good. Only one comment to fix and this can be merged.

File: pkg/networkservice/common/excludedprefixes/utils.go (hunk 3 above)
Comment (Mixaster995) [reply pkg/networkservice/common/excludedprefixes/utils.go]: @denis-tingaikin ok, done

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**